src/main/scala/org/constellation/primitives/ChannelMessage.scala

```diff
@@ -48,37 +54,79 @@ object ChannelMessage {
   }
 
   def createGenesis(
-    channelOpenRequest: ChannelOpenRequest
-  )(implicit dao: DAO): Option[ChannelMessage] = {
-    if (dao.messageService.get(channelOpenRequest.channelId).isEmpty &&
-        !dao.threadSafeMessageMemPool.activeChannels.contains(channelOpenRequest.channelId)) {
-
-      val genesisMessageStr =
-        ChannelOpen(channelOpenRequest.jsonSchema, channelOpenRequest.acceptInvalid).json
-      val msg = create(genesisMessageStr, Genesis.CoinBaseHash, channelOpenRequest.channelId)
-      val semaphore = new Semaphore(1)
-      dao.threadSafeMessageMemPool.activeChannels(channelOpenRequest.channelId) = semaphore
-      semaphore.acquire()
-      dao.threadSafeMessageMemPool.put(Seq(msg), overrideLimit = true)
-      Some(msg)
-    } else None
+    channelOpenRequest: ChannelOpen
+  )(implicit dao: DAO): Future[ChannelOpenResponse] = {
+
+    logger.info(s"Channel open $channelOpenRequest")
+
+    dao.threadSafeMessageMemPool.selfChannelNameToGenesisMessage
+      .get(channelOpenRequest.name)
+      .map { msg =>
+        Future.successful(
+          ChannelOpenResponse("Error: channel name already in use", msg.signedMessageData.hash)
+        )
+      }
+      .getOrElse {
+        val genesisMessageStr = channelOpenRequest.json
+        val msg = create(genesisMessageStr, Genesis.CoinBaseHash, channelOpenRequest.name)
+        dao.threadSafeMessageMemPool.selfChannelNameToGenesisMessage(channelOpenRequest.name) = msg
+        val genesisHashChannelId = msg.signedMessageData.hash
+        dao.threadSafeMessageMemPool.selfChannelIdToName(genesisHashChannelId) =
+          channelOpenRequest.name
+        dao.threadSafeMessageMemPool.put(Seq(msg), overrideLimit = true)
+        val semaphore = new Semaphore(1)
+        dao.threadSafeMessageMemPool.activeChannels(genesisHashChannelId) = semaphore
+        semaphore.acquire()
+        Future {
+          var retries = 0
+          var metadata: Option[ChannelMetadata] = None
+          while (retries < 10 && metadata.isEmpty) {
+            retries += 1
+            Thread.sleep(1000)
+            metadata = dao.channelService.get(genesisHashChannelId)
+          }
+          val response =
+            if (metadata.isEmpty) "Timeout awaiting block acceptance"
+            else {
+              "Success"
+            }
+          ChannelOpenResponse(response, genesisHashChannelId)
```

Comment:
These responses worry be a bit because they seem rather adhoc. Do we intend to document all the possible responses?

Reply:
For right now they're just for displaying in the UI, yes we should create a validation spec for this.